examples/t8code_3d_dgsem/elixir_advection_unstructured_curved.jl

```diff
@@ -0,0 +1,95 @@
+using Downloads: download
+using OrdinaryDiffEq
+using Trixi
+
+###############################################################################
+# semidiscretization of the linear advection equation
+
+advection_velocity = (0.2, -0.7, 0.5)
+equations = LinearScalarAdvectionEquation3D(advection_velocity)
+
+# Create DG solver with polynomial degree = 3 and (local) Lax-Friedrichs/Rusanov flux as surface flux
+solver = DGSEM(polydeg=3, surface_flux=flux_lax_friedrichs)
+
+initial_condition = initial_condition_convergence_test
+
+boundary_condition = BoundaryConditionDirichlet(initial_condition)
+boundary_conditions = Dict(
+  :all => boundary_condition
+)
+
+# Mapping as described in https://arxiv.org/abs/2012.12040 but with less warping.
+# The mapping will be interpolated at tree level, and then refined without changing
+# the geometry interpolant. The original mapping applied to this unstructured mesh
+# causes some Jacobians to be negative, which makes the mesh invalid.
+function mapping(xi, eta, zeta)
+  # Don't transform input variables between -1 and 1 onto [0,3] to obtain curved boundaries
+  # xi = 1.5 * xi_ + 1.5
+  # eta = 1.5 * eta_ + 1.5
+  # zeta = 1.5 * zeta_ + 1.5
+
+  y = eta + 1/6 * (cos(1.5 * pi * (2 * xi - 3)/3) *
+                   cos(0.5 * pi * (2 * eta - 3)/3) *
+                   cos(0.5 * pi * (2 * zeta - 3)/3))
+
+  x = xi + 1/6 * (cos(0.5 * pi * (2 * xi - 3)/3) *
+                  cos(2 * pi * (2 * y - 3)/3) *
+                  cos(0.5 * pi * (2 * zeta - 3)/3))
+
+  z = zeta + 1/6 * (cos(0.5 * pi * (2 * x - 3)/3) *
+                    cos(pi * (2 * y - 3)/3) *
+                    cos(0.5 * pi * (2 * zeta - 3)/3))
+
+  return SVector(x, y, z)
+end
+
+# Unstructured mesh with 68 cells of the cube domain [-1, 1]^3
+mesh_file = joinpath(@__DIR__, "cube_unstructured_1.inp")
+isfile(mesh_file) || download("https://gist.githubusercontent.com/efaulhaber/d45c8ac1e248618885fa7cc31a50ab40/raw/37fba24890ab37cfa49c39eae98b44faf4502882/cube_unstructured_1.inp",
+                              mesh_file)
+
+# INP mesh files are only support by p4est. Hence, we
+# create a p4est connecvity object first from which
+# we can create a t8code mesh.
+conn = Trixi.read_inp_p4est(mesh_file,Val(3))
+
+mesh = T8codeMesh{3}(conn, polydeg=3,
+                    mapping=mapping,
+                    initial_refinement_level=2)
```

Comment:
This also looks like ti may better be another constructor instead of relying on internal code?

Reply:
See discussion below.